### PR TITLE
BFGS GPU Support

### PR DESCRIPTION
**Pull request by avik-pal**
TODO:
- [x] Fails for ArrayPartition type

MWE:
```
using Optim, CUDA
rosenbrock(x) =  (1.0 - x[1])^2 + 100.0 * (x[2] - x[1]^2)^2
result = optimize(rosenbrock, cu(zeros(2)), BFGS())
```

Prev:
```
ERROR: ArgumentError: cannot take the CPU address of a CuArray{Float32, 1}
Stacktrace:
  [1] unsafe_convert(#unused#::Type{Ptr{Float32}}, x::CuArray{Float32, 1})
    @ CUDA ~/.julia/packages/CUDA/mVgLI/src/array.jl:262
  [2] gemv!(trans::Char, alpha::Float32, A::Matrix{Float32}, X::CuArray{Float32, 1}, beta::Float32, Y::CuArray{Float32, 1})
    @ LinearAlgebra.BLAS /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/blas.jl:704
  [3] gemv!(y::CuArray{Float32, 1}, tA::Char, A::Matrix{Float32}, x::CuArray{Float32, 1}, α::Bool, β::Bool)
    @ LinearAlgebra /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/matmul.jl:544
  [4] mul!
    @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/matmul.jl:66 [inlined]
  [5] mul!
    @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/matmul.jl:275 [inlined]
  [6] update_state!(d::OnceDifferentiable{Float32, CuArray{Float32, 1}, CuArray{Float32, 1}}, state::Optim.BFGSState{CuArray{Float32, 1}, Matrix{Float32}, Float32, CuArray{Float32, 1}}, method::BFGS{LineSearches.InitialStatic{Float64}, LineSearches.HagerZhang{Float64, Base.RefValue{Bool}}, Nothing, Nothing, Flat})
    @ Optim ~/testing/Optim.jl/src/multivariate/solvers/first_order/bfgs.jl:119
  [7] optimize(d::OnceDifferentiable{Float32, CuArray{Float32, 1}, CuArray{Float32, 1}}, initial_x::CuArray{Float32, 1}, method::BFGS{LineSearches.InitialStatic{Float64}, LineSearches.HagerZhang{Float64, Base.RefValue{Bool}}, Nothing, Nothing, Flat}, options::Optim.Options{Float64, Nothing}, state::Optim.BFGSState{CuArray{Float32, 1}, Matrix{Float32}, Float32, CuArray{Float32, 1}})
    @ Optim ~/testing/Optim.jl/src/multivariate/optimize/optimize.jl:57
  [8] optimize
    @ ~/testing/Optim.jl/src/multivariate/optimize/optimize.jl:35 [inlined]
  [9] #optimize#87
    @ ~/testing/Optim.jl/src/multivariate/optimize/interface.jl:142 [inlined]
 [10] optimize(f::Function, initial_x::CuArray{Float32, 1}, method::BFGS{LineSearches.InitialStatic{Float64}, LineSearches.HagerZhang{Float64, Base.RefValue{Bool}}, Nothing, Nothing, Flat}, options::Optim.Options{Float64, Nothing}) (repeats 2 times)
    @ Optim ~/testing/Optim.jl/src/multivariate/optimize/interface.jl:141
 [11] top-level scope
    @ REPL[6]:1
```

Curr:
```
    * Status: success

    * Candidate solution
       Final objective value:     1.982578e-05
   
    * Found with
       Algorithm:     BFGS
   
    * Convergence measures
       |x - x'|               = 0.00e+00 ≤ 0.0e+00
       |x - x'|/|x'|          = 0.00e+00 ≤ 0.0e+00
       |f(x) - f(x')|         = 0.00e+00 ≤ 0.0e+00
       |f(x) - f(x')|/|f(x')| = 0.00e+00 ≤ 0.0e+00
       |g(x)|                 = 1.37e-03 ≰ 1.0e-08
   
    * Work counters
       Seconds run:   0  (vs limit Inf)
       Iterations:    47
       f(x) calls:    971
       ∇f(x) calls:   971
```

### CODE REVIEW

**Review by ChrisRackauckas — APPROVED** [review]
You could fuse a bit more operations, but this probably isn't the place for high performance nasty code, so this will give the right mix of generality and conciseness.

File: Project.toml [hunk 1]
```diff
@@ -12,6 +12,7 @@ NaNMath = "77ba4419-2d1f-58cd-9bb1-8ffee604a2e3"
 Parameters = "d96e819e-fc66-5662-9728-84c9c7592b0a"
 PositiveFactorizations = "85a6dd25-e78a-55b7-8502-1745935b8125"
 Printf = "de0858da-6303-5e67-8744-51eddeeeb8d7"
+RecursiveArrayTools = "731186ca-8d62-57ce-b412-fbd966d074cd"
```
Comment (YingboMa) [inline Project.toml:15]: Remember to add a compat

**Comment by pkofod** [comment]
Thanks. I had to approve CI runs since this is your first contribution. Looks good to me.

**Comment by vaishnavtv** [comment]
Hello, I see in the release notes of v1.4 that this pull request has been merged. Does this mean that BFGS now has GPU support? I'm able to use LBFGS, but not BFGS on GPU. 
A MWE: 

```julia
using GalacticOptim, CUDA, Optim, Zygote
CUDA.allowscalar(false)
lossFn(θ, p) = sum(abs, θ)
θ0 = cu(zeros(2));
f = OptimizationFunction(lossFn, GalacticOptim.AutoZygote())
prob = GalacticOptim.OptimizationProblem(f, initθ);
# sol1 = GalacticOptim.solve(prob, Optim.BFGS()); # doesn't work
# sol2 = GalacticOptim.solve(prob, Optim.LBFGS()); # works
```
<details><summary>Details:</summary>
<p>

```julia
ERROR: LoadError: Scalar indexing is disallowed.
Invocation of getindex resulted in scalar indexing of a GPU array.
This is typically caused by calling an iterating implementation of a method.
Such implementations *do not* execute on the GPU, but very slowly on the CPU,
and therefore are only permitted from the REPL for prototyping purposes.
If you did intend to index this array, annotate the caller with @allowscalar.
Stacktrace:
  [1] error(s::String)
    @ Base ./error.jl:33
  [2] assertscalar(op::String)
    @ GPUArrays /scratch/user/vish0908/.julia/packages/GPUArrays/0vqbc/src/host/indexing.jl:53
  [3] getindex(::CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}, ::Int64, ::Int64)
    @ GPUArrays /scratch/user/vish0908/.julia/packages/GPUArrays/0vqbc/src/host/indexing.jl:86
  [4] macro expansion
    @ /scratch/user/vish0908/.julia/packages/Optim/3K7JI/src/multivariate/solvers/first_order/bfgs.jl:166 [inlined]
  [5] macro expansion
    @ ./simdloop.jl:77 [inlined]
  [6] update_h!(d::TwiceDifferentiable{Float32, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}}, state::Optim.BFGSState{CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}, Float32, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}}, method::BFGS{LineSearches.InitialStatic{Float64}, LineSearches.HagerZhang{Float64, Base.RefValue{Bool}}, Nothing, Nothing, Flat})
    @ Optim /scratch/user/vish0908/.julia/packages/Optim/3K7JI/src/multivariate/solvers/first_order/bfgs.jl:165
  [7] optimize(d::TwiceDifferentiable{Float32, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}}, initial_x::CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, method::BFGS{LineSearches.InitialStatic{Float64}, LineSearches.HagerZhang{Float64, Base.RefValue{Bool}}, Nothing, Nothing, Flat}, options::Optim.Options{Float64, GalacticOptim.var"#_cb#103"{GalacticOptim.var"#101#109", BFGS{LineSearches.InitialStatic{Float64}, LineSearches.HagerZhang{Float64, Base.RefValue{Bool}}, Nothing, Nothing, Flat}, Base.Iterators.Cycle{Tuple{GalacticOptim.NullData}}}}, state::Optim.BFGSState{CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}, Float32, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}})
    @ Optim /scratch/user/vish0908/.julia/packages/Optim/3K7JI/src/multivariate/optimize/optimize.jl:71
  [8] optimize(d::TwiceDifferentiable{Float32, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}}, initial_x::CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, method::BFGS{LineSearches.InitialStatic{Float64}, LineSearches.HagerZhang{Float64, Base.RefValue{Bool}}, Nothing, Nothing, Flat}, options::Optim.Options{Float64, GalacticOptim.var"#_cb#103"{GalacticOptim.var"#101#109", BFGS{LineSearches.InitialStatic{Float64}, LineSearches.HagerZhang{Float64, Base.RefValue{Bool}}, Nothing, Nothing, Flat}, Base.Iterators.Cycle{Tuple{GalacticOptim.NullData}}}})
    @ Optim /scratch/user/vish0908/.julia/packages/Optim/3K7JI/src/multivariate/optimize/optimize.jl:35
  [9] __solve(prob::OptimizationProblem{true, OptimizationFunction{true, GalacticOptim.AutoZygote, typeof(lossFn3), Nothing, Nothing, Nothing, Nothing, Nothing, Nothing}, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, SciMLBase.NullParameters, Nothing, Nothing, Nothing, Nothing, Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}}, opt::BFGS{LineSearches.InitialStatic{Float64}, LineSearches.HagerZhang{Float64, Base.RefValue{Bool}}, Nothing, Nothing, Flat}, data::Base.Iterators.Cycle{Tuple{GalacticOptim.NullData}}; maxiters::Nothing, cb::Function, progress::Bool, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ GalacticOptim /scratch/user/vish0908/.julia/packages/GalacticOptim/bEh06/src/solve/optim.jl:55
 [10] __solve (repeats 2 times)
    @ /scratch/user/vish0908/.julia/packages/GalacticOptim/bEh06/src/solve/optim.jl:10 [inlined]
 [11] #solve#476
    @ /scratch/user/vish0908/.julia/packages/SciMLBase/n3U0M/src/solve.jl:3 [inlined]
 [12] solve(::OptimizationProblem{true, OptimizationFunction{true, GalacticOptim.AutoZygote, typeof(lossFn3), Nothing, Nothing, Nothing, Nothing, Nothing, Nothing}, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, SciMLBase.NullParameters, Nothing, Nothing, Nothing, Nothing, Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}}, ::BFGS{LineSearches.InitialStatic{Float64}, LineSearches.HagerZhang{Float64, Base.RefValue{Bool}}, Nothing, Nothing, Flat})
    @ SciMLBase /scratch/user/vish0908/.julia/packages/SciMLBase/n3U0M/src/solve.jl:3
```
</p>
</details>

**Comment by longemen3000** [comment]
The main culprit seems to be this function
https://github.com/JuliaNLSolvers/Optim.jl/blob/01b49c7dc4ec37f97d87118f03877c64e3da9a8c/src/multivariate/solvers/first_order/bfgs.jl#L149

**Pull request merged**